### PR TITLE
Added support for optional callback when scheduled jobs are done

### PR DESCRIPTION
**Pull request by seripap**
I added support for an optional callback to be fired if provided when **job** is completed- just an extension of ``Job``. 

I also wasn't quite sure for the reason of this block:

```javascript
var arg;
for (var i = 0; i < arguments.length; i++) {
   arg = arguments[i];
    if (typeof arg == 'string' || arg instanceof String) {
      name = arg;
    } else {
      this.job = arg;
    }
  }
```

Instead of accessing the arguments obj, I just pass in the parameters to the class directly, to allow for more flexibility and removing the loop. If there was a reason, feel free to stop me and explain the method to the madness.

Cheers.

### CODE REVIEW

**Comment by tejasmanohar** [comment]
Nice work, @seripap! This adds functionality and isn't a breaking change. I also don't see a reason for the block at first glance either but didn't write this part of the code. I don't see a reason not to simply pass arguments to the `Job`.

Before I merge, any thoughts on this? @santigimeno 

I'd like to give the removal of that block another eye or at least some more time before merging, just to be sure.

**Comment by seripap** [comment]
Found a small issue if using the Job api directly with Job(fn, fn). Just made a quick patch.

File: .travis.yml [hunk 1]
```diff
@@ -5,3 +5,4 @@ node_js:
   - "iojs"
 before_script: npm run lint
 script: ./node_modules/.bin/istanbul cover ./node_modules/.bin/nodeunit test && cat ./coverage/lcov.info | ./node_modules/.bin/coveralls && rm -rf ./coverage
+
```
Comment (santigimeno) [inline .travis.yml]: This change seems unrelated

**Comment by seripap** [comment]
I think the coveralls outage is breaking the build, but tests should be passing :heavy_check_mark:

File: .travis.yml (hunk 1 above)
Comment (seripap) [inline .travis.yml]: the build via travis was building so more or less seeing what retriggering would have done. Looks like coveralls is not talking with it.

File: lib/schedule.js [hunk 2]
```diff
@@ -14,31 +14,26 @@ var events = require('events'),
...
+
+  // Make sure callback is actually a callback
+  if (this.job === name) {
+    // Name wasn't provided and maybe a callback is there
+    this.callback = job || false;
+  } else {
+    // Name was provided, and maybe a callback is there
+    this.callback = callback || false;
+  }
+
```
Comment (santigimeno) [inline lib/schedule.js]: May you could check here if callback it's a function or not instead of in prepareNextInvocation

File: test/convenience-method-test.js [hunk 3]
```diff
@@ -237,6 +255,7 @@ module.exports = {
         test.ok(true);
       });
 
+
```
Comment (santigimeno) [inline test/convenience-method-test.js]: remove extra line

**Comment by seripap** [comment]
Thanks for the feedback @santigimeno, the changes are in 05e60de

**Comment by santigimeno** [comment]
@seripap I'm not sure I understand the need of a callback, wouldn't you accomplish the same by calling a function inside the job itself?

**Comment by seripap** [comment]
@santigimeno my use case is purely off of **scheduledJob** running the same function in different intervals that processes different information through a loop. The intervals on run are different per job and If the process didn't finish from the first job and the cron went off again, I wanted it to exit and just wait until the next interval to go off.

```javascript
var schedule = require('node-schedule');

function loop(foo) {
    // something that loops through foo that may take a long time to finish
}

// run every hour
schedule.scheduleJob('0 * * * *', function() {
    if (job1isRunning) {
        return false;
    }

    var bar = [some big array];
    loop(bar);
    job1isRunning = true;
}, function() {
    job1isRunning = false;
})

// run every 15 minutes
schedule.scheduleJob('*/15 * * * *', function() {
    if (job2isRunning) {
        return false;
    }
    var bar2 = [some big array];
    loop(bar2);
    job2isRunning = true;
}, function() {
    job2isRunning = false;
})

// run every 30 minutes
schedule.scheduleJob('*/30 * * * *', function() {
    if (job3isRunning) {
        return false;
    }
    var bar3 = [some big array];
    loop(bar3);
    job3isRunning = true;
}, function() {
    job3isRunning = false;
})
```

**Comment by santigimeno** [comment]
@seripap Thanks for the explanation! I've a question: maybe I'm missing something, but how can a job went off again if the previous one isn't over, being loop function synchronous? In the case it's asynchronous, how can a callback option help?

**Comment by tejasmanohar** [comment]
Travis passed, fyi :)

**Comment by seripap** [comment]
@santigimeno In our use case, It's more of the ability to stop a certain cronjob after it goes off after n minutes and maybe do some more manipulations before restarting the cron.

I based the feature off of [node-cron](https://github.com/ncb000gt/node-cron) which had the ability to fire a callback when a certain cronjob was completed. It could be more useful if the context is actually passed back in the callback.

**Comment by santigimeno** [comment]
@seripap thanks for answering, but I still fail to understand the need of a callback :(. I'm sure I'm missing something obvious: could you please write a very basic example implementing the stop functionality you're referring to?
One more thing: how do you define that a job is done taking into account that a job can be sync or async?

**Comment by seripap** [comment]
@santigimeno sorry i've been out of town.

I'm actually leaving the project that i've implemented this module on but I can tell you we've used the callback for our notification events as well. It was extremely useful in monitoring when cron jobs started and stopped.

It may be useful for another person who manages multiple cron jobs per app, maybe not. It is an optional parameter so it wouldn't hurt if it was added or not.

**Comment by santigimeno** [comment]
@seripap fair enough. Let's see what the others think @tejasmanohar @jonhester ??
I'm -0 on this change

**Comment by iolufemi** [comment]
+1 It will be nice to have this enhancement.

It will be good to be able to set a notification callback when a cronjob runs, especially when using this module on delicate systems like Payment systems.

**Comment by tejasmanohar** [comment]
Yeah, this makes to me. It's a harmless enhancement.
